tiledb/sm/rtree/rtree.h

```diff
@@ -0,0 +1,206 @@
+
+#include <memory>
+#include <vector>
+#include "tiledb/sm/buffer/buffer.h"
+#include "tiledb/sm/buffer/const_buffer.h"
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/enums/layout.h"
+#include "tiledb/sm/misc/status.h"
+#include "tiledb/sm/misc/tile_overlap.h"
+#include "tiledb/sm/subarray/subarray.h"
+
+namespace tiledb {
+namespace sm {
+
+/**
+ * A simple RTree implementation. It supports storing only n-dimensional
+ * MBRs (not points). Also it only offers bottom-up bulk-loading
+ * (without incremental updates), and range and point queries.
+ */
+class RTree {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /** Constructor. */
+  RTree();
+
+  /**
+   * Constructor. This admits a list of sorted MBRs that will
+   * constitute the leaf level of the tree. The constructor will
+   * construct bottom up the tree based on these ``mbrs``.
+   * The input MBRs will be copied into the leaf level.
+   */
+  RTree(
+      Datatype type,
+      unsigned dim_num,
+      unsigned fanout,
+      const std::vector<void*>& mbrs);
+
+  /** Destructor. */
+  ~RTree();
+
+  /** Copy constructor. This performs a deep copy. */
+  RTree(const RTree& rtree);
+
+  /** Move constructor. */
+  RTree(RTree&& rtree) noexcept;
+
+  /** Copy-assign operator. This performs a deep copy. */
+  RTree& operator=(const RTree& rtree);
+
+  /** Move-assign operator. */
+  RTree& operator=(RTree&& rtree) noexcept;
+
+  /* ********************************* */
+  /*                 API               */
+  /* ********************************* */
+
+  /** Returns the number of dimensions. */
+  unsigned dim_num() const;
+
+  /** Returns the fanout. */
+  unsigned fanout() const;
+
+  /**
+   * Returns the tile overlap of the input range with the MBRs stored
+   * in the RTree.
+   */
+  template <class T>
+  TileOverlap get_tile_overlap(const std::vector<const T*>& range) const;
+
+  /** Returns the tree height. */
+  unsigned height() const;
+
+  /**
+   * Returns the overlap between a range and an RTree MBR, as the ratio
+   * of the volume of the overlap over the volume of the MBR.
+   */
+  template <class T>
+  double range_overlap(const std::vector<const T*>& range, const T* mbr) const;
+
+  /**
+   * Returns the number of leaves that are stored in a (full) subtree
+   * rooted at the input level. Note that the root is at level 0.
+   */
+  uint64_t subtree_leaf_num(uint64_t level) const;
+
+  /** Returns the datatype of the R-Tree. */
+  Datatype type() const;
+
+  /** Serializes the contents of the object to the input buffer. */
+  Status serialize(Buffer* buff) const;
+
+  /** Deserializes the contents of the object from the input buffer. */
+  Status deserialize(ConstBuffer* cbuff);
+
+ private:
+  /* ********************************* */
+  /*      PRIVATE TYPE DEFINITIONS     */
+  /* ********************************* */
+
+  /** A level of the RTree. */
```

Comment:
These two structs `Level` and `Entry` deserve slightly more detail in their doc comments. Just something describing the representation used here.